File: python_code/line/algo/linkedlist/_203_RemoveElements.py
from line.algo.linkedlist.ListNode import ListNode


class Solution:
    def removeElements(self, head: ListNode, val: int) -> ListNode:
        dummy_node = ListNode(-1)
        dummy_node.next = head

        prev, curr = dummy_node, head
        while curr:
            if curr.val == val:
                prev.next = curr.next
            else:
                prev = curr
            curr = prev.next

        return dummy_node.next

File: cpp_code/line/algo/linkedlist/_203_RemoveElements.cpp
/**
* 抖码算法，让算法学习变的简单有趣
* @作者 : 老汤
*/

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode* dummyNode = new ListNode(-1);
        dummyNode->next = head;

        ListNode *curr = head;
        ListNode *prev = dummyNode;
        while (curr != nullptr) {
            if (curr->val == val) {
                prev->next = curr->next;
                curr->next = nullptr;
            } else {
                prev = curr;
            }
            curr = prev->next;
        }

        return dummyNode->next;
    }
};